src/libPMacc/include/eventSystem/events/CudaEventHandle.hpp

```diff
@@ -0,0 +1,132 @@
+/**
+ * Copyright 2014-2016 Rene Widera
+ *
+ * This file is part of libPMacc.
+ *
+ * libPMacc is free software: you can redistribute it and/or modify
+ * it under the terms of either the GNU General Public License or
+ * the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * libPMacc is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License and the GNU Lesser General Public License
+ * for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * and the GNU Lesser General Public License along with libPMacc.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+
+#pragma once
+
+#include "pmacc_types.hpp"
+#include <cuda_runtime.h>
+#include "eventSystem/events/CudaEvent.def"
+
+namespace PMacc
+{
+
+/** handle to CudaEvent */
+class CudaEventHandle
+{
+private:
+
+    /** pointer to the CudaEvent */
+    CudaEvent* event;
+
+public:
+
+    /** create invalid handle  */
+    CudaEventHandle( ) : event( NULL )
+    {
+
+    }
+
+    /** create a handle to a valid CudaEvent
+     *
+     * @param evPointer pointer to a CudaEvent
+     */
+    CudaEventHandle( CudaEvent* const evPointer ) : event( evPointer )
+    {
+        event->registerHandle();
+    }
+
+    CudaEventHandle( const CudaEventHandle& other ) : event( NULL )
+    {
+        /* register and release handle is done by the assign operator */
+        *this = other;
+    }
+
+    /** assign an event handle
+     *
+     * undefined behavior if the other event handle is equal to this instance
+     *
+     * @param other event handle
+     * @return this handle
+     */
+    CudaEventHandle&
+    operator=( const CudaEventHandle& other )
+    {
+        /* check if an old event is overwritten */
+        if( event )
+            event->releaseHandle( );
+        event = other.event;
+        event->registerHandle( );
+        return *this;
+    }
+
+    /** Destructor */
+    ~CudaEventHandle( )
+    {
+        if( event )
+            event->releaseHandle( );
+        event = NULL;
+    }
+
+    /**
+     * get native cuda event
+     *
+     * @return native cuda event
+     */
+    cudaEvent_t operator*( ) const
+    {
+        assert( event );
+        return **event;
```

Comment:
that looks super-odd but is indeed valid since the CudaEvent wrapper is always valid/RAII